sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java

```diff
@@ -111,20 +115,39 @@ private File downloadFile(String inputPath, String mimeType) throws IOException
     }
   }
 
+  private File getLocalJar(String inputJarPath) throws IOException {
+    if (!jarCache.containsKey(inputJarPath)) {
+      jarCache.put(inputJarPath, downloadFile(inputJarPath, "application/java-archive"));
+    }
+    return jarCache.get(inputJarPath);
+  }
+
   private ClassLoader createClassLoader(String inputJarPath) throws IOException {
-    File tmpJar = downloadFile(inputJarPath, "application/java-archive");
+    File tmpJar = getLocalJar(inputJarPath);
     return new URLClassLoader(new URL[] {tmpJar.toURI().toURL()});
   }
 
+  public ClassLoader createClassLoader(List<String> inputJarPaths) throws IOException {
+    List<File> localJars = new ArrayList<>();
+    for (String inputJar : inputJarPaths) {
+      localJars.add(getLocalJar(inputJar));
```

Comment:
Couldn't you merge these into a single loop?
`urls.add(getLocalJar(inputJar).toURI().toURL())`

Reply:
Yep, done. (I originally tried to do this the functional way using streams, but Java doesn't really let you throw checked exceptions from lambdas. 😢 )